### ISSUE TITLE
Can't build on Linux

### ISSUE THREAD

**Issue by darealshinji** [Senscape/Dagon#174]
The latest code doesn't build on Linux. I've tried it on Ubuntu 16.04 using g++ 5.4 as well as Ubuntu 14.04 (chroot) using g++ 4.8. Even adding `-std=gnu++11` or `-std=c++11` to the build command doesn't help.

```
djcj ~/Downloads/Dagon
$ premake4 gmake
WARNING: Library stdc++ not found
Building configurations...
Running action 'gmake'...
Generating build/Makefile...
Generating build/Dagon.make...
Done.
djcj ~/Downloads/Dagon
$ make -C build/
make: Entering directory '/home/djcj/Downloads/Dagon/build'
==== Building Dagon (debug) ====
Creating debug
Creating objs/debug
ShaderData.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
stb_image.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
SplashData.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
siphash24.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
FontData.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
dirent.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
DustData.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
Group.cpp
Deserializer.cpp
In file included from ../src/AudioAsset.h:17:0,
                 from ../src/Audio.h:39,
                 from ../src/Deserializer.cpp:19:
../src/Asset.h:56:32: error: 'shared_ptr' in namespace 'std' does not name a template type
   size_t operator()(const std::shared_ptr<Asset>& ptr) const {
                                ^
../src/Asset.h:56:42: error: expected ',' or '...' before '<' token
   size_t operator()(const std::shared_ptr<Asset>& ptr) const {
                                          ^
../src/Asset.h: In member function 'size_t dagon::AssetPtrHash::operator()(int) const':
../src/Asset.h:57:37: error: 'ptr' was not declared in this scope
     return std::hash<std::string>()(ptr->id());
                                     ^
In file included from ../src/Deserializer.cpp:19:0:
../src/Audio.h: At global scope:
../src/Audio.h:115:8: error: 'shared_ptr' in namespace 'std' does not name a template type
   std::shared_ptr<AudioAsset> _asset;
        ^
In file included from ../src/Deserializer.cpp:20:0:
../src/AudioManager.h:64:33: error: 'weak_ptr' is not a member of 'std'
   std::unordered_map<AssetID_t, std::weak_ptr<AudioAsset>> _activeAssets;
                                 ^
../src/AudioManager.h:64:33: error: 'weak_ptr' is not a member of 'std'
../src/AudioManager.h:64:47: error: template argument 2 is invalid
   std::unordered_map<AssetID_t, std::weak_ptr<AudioAsset>> _activeAssets;
                                               ^
../src/AudioManager.h:64:47: error: template argument 5 is invalid
../src/AudioManager.h:64:57: error: expected unqualified-id before '>' token
   std::unordered_map<AssetID_t, std::weak_ptr<AudioAsset>> _activeAssets;
                                                         ^
../src/AudioManager.h:88:8: error: 'shared_ptr' in namespace 'std' does not name a template type
   std::shared_ptr<AudioAsset> asAsset(const AssetID_t& id);
        ^
In file included from /usr/include/c++/5/bits/hashtable.h:35:0,
                 from /usr/include/c++/5/unordered_map:47,
                 from ../src/AudioManager.h:37,
                 from ../src/Deserializer.cpp:20:
/usr/include/c++/5/bits/hashtable_policy.h: In instantiation of 'struct std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash>':
/usr/include/c++/5/type_traits:137:12:   required from 'struct std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> >'
/usr/include/c++/5/type_traits:148:38:   required from 'struct std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
/usr/include/c++/5/bits/unordered_set.h:95:63:   required from 'class std::unordered_set<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash>'
../src/Room.h:90:60:   required from here
/usr/include/c++/5/bits/hashtable_policy.h:85:34: error: no match for call to '(const dagon::AssetPtrHash) (const std::shared_ptr<dagon::Asset>&)'
  noexcept(declval<const _Hash&>()(declval<const _Key&>()))>
                                  ^
In file included from ../src/AudioAsset.h:17:0,
                 from ../src/Audio.h:39,
                 from ../src/Deserializer.cpp:19:
../src/Asset.h:56:10: note: candidate: size_t dagon::AssetPtrHash::operator()(int) const
   size_t operator()(const std::shared_ptr<Asset>& ptr) const {
          ^
../src/Asset.h:56:10: note:   no known conversion for argument 1 from 'const std::shared_ptr<dagon::Asset>' to 'int'
In file included from /usr/include/c++/5/bits/move.h:57:0,
                 from /usr/include/c++/5/bits/stl_pair.h:59,
                 from /usr/include/c++/5/bits/stl_algobase.h:64,
                 from /usr/include/c++/5/bits/char_traits.h:39,
                 from /usr/include/c++/5/string:40,
                 from ../src/Node.h:21,
                 from ../src/Deserializer.h:21,
                 from ../src/Deserializer.cpp:18:
/usr/include/c++/5/type_traits: In instantiation of 'struct std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >':
/usr/include/c++/5/bits/unordered_set.h:95:63:   required from 'class std::unordered_set<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash>'
../src/Room.h:90:60:   required from here
/usr/include/c++/5/type_traits:148:38: error: 'value' is not a member of 'std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> >'
     : public integral_constant<bool, !_Pp::value>
                                      ^
In file included from /usr/include/c++/5/unordered_set:48:0,
                 from ../src/Room.h:25,
                 from ../src/Deserializer.cpp:25:
/usr/include/c++/5/bits/unordered_set.h: In instantiation of 'class std::unordered_set<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash>':
../src/Room.h:90:60:   required from here
/usr/include/c++/5/bits/unordered_set.h:95:63: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef __uset_hashtable<_Value, _Hash, _Pred, _Alloc>  _Hashtable;
                                                               ^
/usr/include/c++/5/bits/unordered_set.h:102:45: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::key_type key_type;
                                             ^
/usr/include/c++/5/bits/unordered_set.h:103:47: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::value_type value_type;
                                               ^
/usr/include/c++/5/bits/unordered_set.h:104:43: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::hasher hasher;
                                           ^
/usr/include/c++/5/bits/unordered_set.h:105:46: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::key_equal key_equal;
                                              ^
/usr/include/c++/5/bits/unordered_set.h:106:51: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::allocator_type allocator_type;
                                                   ^
/usr/include/c++/5/bits/unordered_set.h:111:45: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::pointer  pointer;
                                             ^
/usr/include/c++/5/bits/unordered_set.h:112:50: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::const_pointer const_pointer;
                                                  ^
/usr/include/c++/5/bits/unordered_set.h:113:47: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::reference  reference;
                                               ^
/usr/include/c++/5/bits/unordered_set.h:114:52: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::const_reference const_reference;
                                                    ^
/usr/include/c++/5/bits/unordered_set.h:115:46: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::iterator  iterator;
                                              ^
/usr/include/c++/5/bits/unordered_set.h:116:51: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::const_iterator const_iterator;
                                                   ^
/usr/include/c++/5/bits/unordered_set.h:117:51: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::local_iterator local_iterator;
                                                   ^
/usr/include/c++/5/bits/unordered_set.h:118:57: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::const_local_iterator const_local_iterator;
                                                         ^
/usr/include/c++/5/bits/unordered_set.h:119:47: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::size_type  size_type;
                                               ^
/usr/include/c++/5/bits/unordered_set.h:120:52: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       typedef typename _Hashtable::difference_type difference_type;
                                                    ^
/usr/include/c++/5/bits/unordered_set.h:274:7: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       operator=(initializer_list<value_type> __l)
       ^
/usr/include/c++/5/bits/unordered_set.h:368:2: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
  emplace(_Args&&... __args)
  ^
/usr/include/c++/5/bits/unordered_set.h:412:7: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       insert(const value_type& __x)
       ^
/usr/include/c++/5/bits/unordered_set.h:416:7: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       insert(value_type&& __x)
       ^
/usr/include/c++/5/bits/unordered_set.h:471:7: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       insert(initializer_list<value_type> __l)
       ^
/usr/include/c++/5/bits/unordered_set.h:616:7: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       equal_range(const key_type& __x)
       ^
/usr/include/c++/5/bits/unordered_set.h:620:7: error: 'value' is not a member of 'std::__not_<std::__and_<std::__is_fast_hash<dagon::AssetPtrHash>, std::__detail::__is_noexcept_hash<std::shared_ptr<dagon::Asset>, dagon::AssetPtrHash> > >'
       equal_range(const key_type& __x) const
       ^
../src/Deserializer.cpp: In member function 'bool dagon::Deserializer::readHeader()':
../src/Deserializer.cpp:87:9: error: 'memcmp' is not a member of 'std'
     if (std::memcmp(SAVIdent, buf, sizeof(SAVIdent)) != 0) {
         ^
../src/Deserializer.cpp:87:9: note: suggested alternative:
In file included from /usr/include/SDL2/SDL_stdinc.h:60:0,
                 from /usr/include/SDL2/SDL_rwops.h:32,
                 from ../src/Deserializer.h:24,
                 from ../src/Deserializer.cpp:18:
/usr/include/string.h:65:12: note:   'memcmp'
 extern int memcmp (const void *__s1, const void *__s2, size_t __n)
            ^
Dagon.make:197: recipe for target 'objs/debug/Deserializer.o' failed
make[1]: *** [objs/debug/Deserializer.o] Error 1
Makefile:16: recipe for target 'Dagon' failed
make: *** [Dagon] Error 2
make: Leaving directory '/home/djcj/Downloads/Dagon/build'
djcj ~/Downloads/Dagon
$ 

```

**Comment by bd339** [comment]
Those look like the same errors I got when trying to resolve #172 . I don't know if MSVC include headers differently than GCC. Apparently it does. I got past those relatively easily by including headers as indicated by [cppreference](cppreference.com). GCC 5.4 should be able to compile the code without `-std=c++11` so I'll send a PR to fix these errors. Let me know here if I don't get all of them (don't have access to a Linux system right now).
